app/workers/search/index_worker.rb

```diff
@@ -0,0 +1,18 @@
+module Search
+  class IndexWorker
+    include Sidekiq::Worker
+
+    sidekiq_options queue: :default, retry: 10
+
+    def perform(record_type, record_id)
+      raise InvalidRecordType unless %w[Comment Article User PageView].include?(record_type)
+
+      record = record_type.constantize.find_by(id: record_id)
```

Comment:
Since we already manually validate the constants we could switch to [safe_constantize](https://api.rubyonrails.org/classes/ActiveSupport/Inflector.html#method-i-safe_constantize) here. Alternatively we could remove the custom validation above and just rely on the exception raised by `constantize`, which actually includes the wrong class name, which would be more helpful for debugging.

```ruby
NameError: uninitialized constant Usre
```

Or if we want to combine the approaches:

```ruby
def perform(record_type, record_id)
  record_class = record_type.safe_constantize
  raise InvalidRecordType, "Invalid class: #{record_type}" unless record_class
  record = record_class.find_by(id: record_id)
  return unless record

  record.index!
end
```

Reply:
I like this - great point! Was the custom validation there originally because this `IndexWorker` is called by different record types (i.e. `User`, `Comment`, etc.) that we index? And we wanted to make sure we only called `.index!` for record types where it's setup?

In other words, if we remove the custom validation and in the future, could someone then call `IndexWorker.perform("Notification", 1)`? Which would fail since `index!` isn't set up for `Notification`? Or is it okay to leave that avenue open since (hopefully) the error message would still be helpful saying `.index!` doesn't work on `Notification`?

Reply:
Good point, it doesn't only need to be an existing model, but also an indexable one. So maybe just leave it a moving the array of valid classes out to an Array, and maybe an a message to the custom exception when we raise it.

Reply:
I wouldn't just rely on `safe_constantize` here because it won't make sure that we really intend to constantize and run the logic on this object, just check that the class exists in the codebase. I think it's better to be confident about what exactly are we doing in this case, not allow passing all the possible classes.

Reply:
Yes, we reached this same conclusion as you @lightalloy:

> it doesn't only need to be an existing model, but also an indexable one